core/integration/ccdb/plugin.go

```diff
@@ -0,0 +1,296 @@
+		return "", fmt.Errorf("could not create a command for CCDB interface because run number is 0")
+	}
+	cmd += " -r " + strconv.FormatUint(uint64(grp.runNumber), 10)
+	if grp.hbfPerTf != 0 {
+		cmd += " -n " + strconv.FormatUint(uint64(grp.hbfPerTf), 10)
+	}
+	// TODO enable once PR#8082 to O2 is at P2
+	//if grp.runType != runtype.NONE {
+	//	cmd += " --run-type " + grp.runType.String()
+	//}
+	if len(grp.detectors) != 0 {
+		cmd += " -d \"" + strings.Join(grp.detectors, ",") + "\""
+	}
+	if len(grp.continuousReadoutDetectors) != 0 {
+		cmd += " -c \"" + strings.Join(grp.continuousReadoutDetectors, ",") + "\""
+	}
+	if len(grp.triggeringDetectors) != 0 {
+		// keeping the full argument until the conflict between '-t's of triggering and run-type is resolved
+		cmd += " --triggering \"" + strings.Join(grp.triggeringDetectors, ",") + "\""
+	}
+	if len(grp.startTimeMs) > 0 {
+		cmd += " -s " + grp.startTimeMs
+	}
+	if len(grp.endTimeMs) > 0 {
+		cmd += " -e " + grp.endTimeMs
+	}
+
+	cmd += " --ccdb-server " + ccdbUrl
+	return
+}
+
+func (p *Plugin) CallStack(data interface{}) (stack map[string]interface{}) {
+	call, ok := data.(*callable.Call)
+	if !ok {
+		return
+	}
+	varStack := call.VarStack
+	envId, ok := varStack["environment_id"]
+	if !ok {
+		log.Error("cannot acquire environment ID")
+		return
+	}
+
+	stack = make(map[string]interface{})
+	stack["RunStart"] = func() (out string) { // must formally return string even when we return nothing
+
+		log.WithField("partition", envId).Debug("performing CCDB interface Run Start")
+		return p.uploadCurrentGRP(varStack, envId)
+	}
+	stack["RunStop"] = func() (out string) {
+		log.WithField("partition", envId).Debug("performing CCDB interface Run Stop")
+		return p.uploadCurrentGRP(varStack, envId)
+	}
+
+	return
+}
+
+func (p *Plugin) uploadCurrentGRP(varStack map[string]string, envId string) string {
+	grp := NewGRPObject(varStack)
+
+	if grp == nil {
+		return fmt.Sprintf("Failed to create a GRP object")
+	}
+	log.WithField("partition", envId).Debug(
+		fmt.Sprintf("GRP: %d, %s, %s, %s, %d, %s, %s, %s, %s",
+			grp.runNumber, grp.runType.String(), grp.startTimeMs, grp.endTimeMs, grp.hbfPerTf, grp.lhcPeriod,
+			strings.Join(grp.detectors, ","), strings.Join(grp.triggeringDetectors, ","), strings.Join(grp.continuousReadoutDetectors, ",")))
+	cmdStr, err := p.NewCcdbGrpWriteCommand(grp, p.ccdbUrl)
+	if err != nil {
+		return fmt.Sprintf("Failed to build a GRP to CCDB upload command: " + err.Error())
+	}
+	log.WithField("partition", envId).Debug(fmt.Sprintf("CCDB GRP upload command: '" + cmdStr + "'"))
+
+	cmd := exec.Command("bash", "-c", cmdStr)
+	// execute the DPL command in the repo of the workflow used
+	cmd.Dir = "/tmp"
+	var cmdOut []byte
+	cmdOut, err = cmd.CombinedOutput()
+	log.Debug("CCDB GRP upload command out: " + string(cmdOut))
+	if err != nil {
+		return fmt.Sprintf("Failed to run the command to upload GRP to CCDB: " + err.Error() + "\ncommand out : " + string(cmdOut))
+	}
```

Comment:
This all looks correct and very good, just one concern: can the command block? If yes, we need a timeout.

Reply:
If we make a plugin non-critical, can it still block?
I will add a timeout just in case though.

Reply:
It can. `critical` only refers to whether a failure stops the env or not, regardless of whether the failure is a timeout or some other kind of failure mode.